tlog-mirror.md

```diff
@@ -0,0 +1,297 @@
+
+The mirror update process is designed to be safely interruptible, while avoding
+large atomic operations. For each origin log, a mirror maintains the following:
+
+* A copy of the log. The latest checkpoint of this log copy is known as the
+  *mirror checkpoint*.
+
+* A *pending checkpoint*, which is at or ahead of the mirror checkpoint. If
+  ahead of the mirror checkpoint, the pending checkpoint describes entries that
+  have yet to be incorporated into the mirror.
+
+* A list of *pending entries* that have yet to be incorporated into the mirror
+  checkpoint. The mirror's *next entry* is the log index of the first entry that
+  is not in either the log or pending entry list.
+
+The update process ensures that all current and past pending checkpoints are
+consistent, and all pending entries are contained in the current pending
+checkpoint. Thus mirrors MAY commit pending entries to the log, serving them as
+entry bundles, as soon as they are added. That is, a mirror MAY use the same
+underlying storage for entry bundles and pending entries, without distinguishing
+between them. It is expected that most tiled log implementations will do this.
+
+Mirrors update in three stages:
+
+1. A mirror client updates the pending checkpoint with a signed checkpoint and a
+   consistency proof.
+
+2. A mirror client uploads new entries to the pending entry list, up to the
+   pending checkpoint.
+
+3. The mirror commits the pending entries to the log and updates the mirror
+   checkpoint.
+
+The next sections describe the HTTP endpoints used by mirror clients to update
+the log.
+
+### add-checkpoint
+
+The mirror implements a [witness][]'s `add-checkpoint` endpoint to update its
+pending checkpoint for a log:
+
+    POST <submission prefix>/add-checkpoint
+
+The request is handled identically to that of a witness, updating the pending
+checkpoint (but not the mirror checkpoint), with the exception that it does not
+need to generate and respond with any cosignatures. The mirror MAY handle the
+request by internally updating the pending checkpoint and responding with an
+empty response body. The mirror MUST retain the log's signature in the pending
+checkpoint.
+
+The mirror cosigner MUST NOT sign the checkpoint in this process. It MAY respond
+with witness cosignatures if the mirror operator wishes to additionally provide
+a separate witness service using its pending checkpoint. If so, this witness
+service MUST be a distinct cosigner from the mirror cosigner, with a distinct
+name. The mirror's signature is computed later, as described below.
+
+### add-entries
+
+The mirror implements an `add-entries` endpoint to upload entries for a supported
+log:
+
+    POST <submission prefix>/add-entries
+
+#### Request Body
+
+The request body MUST have `Content-Type` of `application/octet-stream` and
+contain the following values, concatenated.
+
+* 2 bytes, encoding a big-endian uint16: `log_origin_size`
+* `log_origin_size` bytes, containing the log origin
```

Comment:
We use the term `log_origin` below, so maybe just define it explicitly here in case someone is doing a text search through the document.
```suggestion
* `log_origin_size` bytes, containing the log origin: `log_origin`
```